### PR TITLE
HPCC-19879 Convert SEQUENTIAL to ORDERED inside a child query

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by ghalliday** [comment]
@shamser please review
This should be documented in the red book

File: ecl/hqlcpp/hqlttcpp.cpp [hunk 1]
```diff
@@ -1452,17 +1452,20 @@ IHqlExpression * SequenceNumberAllocator::createTransformed(IHqlExpression * exp
...
+                if (!cur->isConstant())
                 {
-                    if (!next->isConstant())
-                        next.setown(createValue(no_evaluate_stmt, makeVoidType(), next.getClear()));
-                    else
-                        next.clear();
-                }
+                    if (!cur->isAction())
+                    {
+                        HqlExprArray args;
```
Comment (shamser) [inline ecl/hqlcpp/hqlttcpp.cpp]: This 'args' variable shadows the 'args' in line 1452.  Potential to cause confusion/problems.  May be rename.

File: ecl/hqlcpp/hqlttcpp.cpp [hunk 2]
```diff
@@ -1452,17 +1452,20 @@ IHqlExpression * SequenceNumberAllocator::createTransformed(IHqlExpression * exp
...
+                        cur->unwindList(args, no_compound);
+                        unsigned last = args.ordinality()-1;
+                        args.replace(*createValue(no_evaluate_stmt, makeVoidType(), &OLINK(args.item(last))), last);
```
Comment (shamser) [inline ecl/hqlcpp/hqlttcpp.cpp]: What is it doing here?  A comment would be useful here.

File: ecl/hqlcpp/hqlttcpp.cpp [hunk 3]
```diff
@@ -13709,6 +13804,13 @@ void HqlCppTranslator::normalizeGraphForGeneration(HqlExprArray & exprs, HqlQuer
     allocateSequenceNumbers(exprs);                                             // Added to all expressions/output statements etc.
 
     traceExpressions("allocate Sequence", exprs);
+
+    if (options.transformNestedSequential)
+    {
+        transformNestedSequential(exprs);
+        traceExpressions("allocate Sequence", exprs);
```
Comment (shamser) [inline ecl/hqlcpp/hqlttcpp.cpp]: traceExpressions label incorrect ("allocate Sequence") - label should be renamed

File: testing/regress/ecl/redissynctest.ecl [hunk 4]
```diff
@@ -177,7 +177,7 @@ END;
 pubDS := DATASET(N2, TRANSFORM({ integer a }, SELF.a := pub('PubSubTest' + (STRING)COUNTER)));
 
 INTEGER pub2(STRING channel) := FUNCTION
-        sl := SEQUENTIAL(
+        sl := ORDERED(
```
Comment (shamser) [inline testing/regress/ecl/redissynctest.ecl]: This change does not seem directly related to this issue.

**Review by shamser — CHANGES_REQUESTED** [review]
@ghalliday  There are some comments.  Is there a regression query for this issue?

File: testing/regress/ecl/redissynctest.ecl (hunk 4 above)
Comment (ghalliday) [reply testing/regress/ecl/redissynctest.ecl]: This is to avoid the new warning generated when running the regression suite.

**Comment by ghalliday** [comment]
@shamser see changes

**Review by shamser — APPROVED** [review]
@richardkchapman Code reviewed.  Looks good.

**Comment by richardkchapman** [comment]
@ghalliday Please squash

**Comment by ghalliday** [comment]
@richardkchapman squashed

**Pull request merged**